### ISSUE TITLE
[BUG] Volume Expansion Fails When The Larger Replica Doesn't Have Room to Expand

### ISSUE THREAD

**Issue by jdbaudean** [longhorn/longhorn#8513]
## Describe the bug

When I try to expand a volume it will fail if one of the replicas is on a node without enough free space to schedule the volume with the larger size.  Example:

Current volume size 20G

Desired Volume size 50G

One replica is on a node with only 20G left until it becomes unschedulable (or it's already unschedulable).
Instead of rebuilding the replica on a node with enough free space, the resize just fails.  Shouldn't Longhorn be smart enough to move/rebuild the replica on a node with enough free space?

## To Reproduce

Create a volume.  Try to expand that volume to a larger size than is available on one of the nodes where one of the replicas reside.  

## Expected behavior

The replica will rebuild on a node that has enough room for the expanded size.

## Support bundle for troubleshooting

None available
## Environment

 - Longhorn version: 1.5.3
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): Rancher
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: RKE2
   - Number of control plane nodes in the cluster: 3
   - Number of worker nodes in the cluster: 60+
 - Node config
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): AWS
 - Number of Longhorn volumes in the cluster: 260+

## Additional context

This has been happening since at least v1.3.x

**Comment by PhanLe1010** [comment]
Thanks @jdbaudean for reporting.

This ticket looks like a feature request instead of a bug.

We will discuss with other team members

**Comment by PhanLe1010** [comment]
IMO, For automatically rebuild the replica on new nodes if expansion failed, while I think it certainly has value, it is also a very costly behavior. The reason is that rebuilding is an expensive operation. 

Together with the fact that users can already see the expansion error easily and fix it by doing replica eviction or make more space for full nodes, I am more inclined to not implementing this feature.

WDYT @derekbit @shuo-wu @ejweber

**Comment by hoo29** [comment]
If it helps with justification, we recently increased storage across 80+ disks and whilst we had plenty of total storage available on nodes, we had to play whack-a-mole with disk eviction to get everything expanded. It took considerable time. We kept hitting 
https://github.com/longhorn/longhorn/issues/8450 as well which made the situation worse and required more evictions. 

Probably one for a the harvester repo but the the status message gets stuck on expanding, you have to open longhorn to find out why it is stuck.

**Comment by jdbaudean** [comment]
I've run into the same problem as @hoo29, it turns into a game of evicting replicas until you have space to expand.  If there was a way to move/rebuild replicas on a specific node that would make it much easier.  

I'll give a little more details for my use case if that helps.  We run an app in our cluster called coder (https://coder.com/), where people create a workspace (just a docker container really) with a persistent home directory, and we use Longhorn for that persistent volume.  If a user wants to make their workspace larger it might work or it might not, just depending on where their volume replicas are located.  If it doesn't work, we have to hunt down where the volume replicas are, find the ones that don't have enough space to rebuild with the larger size, and then delete them and hope they start rebuilding on a node with sufficient storage.  Either that or tell people to save off all of their data some place else and then delete and recreate their workspace with the larger size.

I know we could technically disable scheduling on all nodes except for the ones with enough space to allow the volume to expand, but that's not feasible when you have hundreds of users constantly creating, attaching, and detaching volumes.

**Comment by hoo29** [comment]
@jdbaudean out of curiosity do you hit https://github.com/longhorn/longhorn/issues/8450 at all where disks on a node get over provisioned?

**Comment by jdbaudean** [comment]
@hoo29 Yes now that you mention it I do see that.  We allow for 150% over provisioning, our nodes have a single 2TB ebs volume attached to them for Longhorn with 30% reserved, leaving 1.37T available, 150% puts that at 2098.97Gi.  I see one at 2311/2098 Gi and another at 2399/2098 Gi.  Also, the "Actual Size" of volumes routinely grow much larger than the "Size".  I'm looking at one volume now that was created as a 50G volume but the actual size is 81G.

**Comment by PhanLe1010** [comment]
I see. Thanks @jdbaudean and @hoo29 for providing additional user store for the feature request. Putting this ticket into the backlog for @derekbit to arrange correct assignee and milestone

**Comment by PhanLe1010** [comment]
> I see one at 2311/2098 Gi and another at 2399/2098 Gi.

I agree this is the bug in the ticket https://github.com/longhorn/longhorn/issues/8450

> Also, the "Actual Size" of volumes routinely grow much larger than the "Size". I'm looking at one volume now that was created as a 50G volume but the actual size is 81G.

For this one, it sounds like a normal behavior when the volume has multiple snapshots. The actual size can be bigger than spec size in this case. Could you please check the doc at https://longhorn.io/docs/1.6.1/nodes-and-volumes/volumes/volume-size/

**Comment by Kudze** [comment]
Just wanna bump this issue, because I have also experienced it.

One thing to note would be that as is its pretty annoying, that its not possible to move "replica" to another node, there is already an feature request issue about it #1400. Had to disable volume scheduling on "almost full" node, so that the replica would go to another node.
